### PR TITLE
Improves OptimizationSystem

### PR DESCRIPTION
**Pull request by ValentinKaisermayer**

### CODE REVIEW

**Comment by ValentinKaisermayer** [comment]
closes #1785

File: src/systems/optimization/optimizationsystem.jl [hunk 1]
```diff
@@ -26,25 +26,22 @@ struct OptimizationSystem <: AbstractTimeIndependentSystem
...
+    The default values to use when initial guess and/or
+    parameters are not supplied in `OptimizationProblem`.
     """
     defaults::Dict
     function OptimizationSystem(op, states, ps, var_to_name, observed,
                                 constraints, name, systems, defaults;
                                 checks::Bool = true)
         if checks
-            check_units(op)
+            op isa Num && check_units(op)
```
Comment (ValentinKaisermayer) [inline src/systems/optimization/optimizationsystem.jl:44]: I'm not sure about this. The intended use-case is when one system has no objective function i.e. just `0` but sub-systems do have one.

Comment (YingboMa) [reply src/systems/optimization/optimizationsystem.jl:44]: MTK should never work with Num. You should always unwrap it and then check if it's Symbolic.

**Review by YingboMa — APPROVED** [review]

File: src/systems/abstractsystem.jl [hunk 2]
```diff
@@ -179,8 +179,6 @@ for prop in [:eqs
              :systems
              :structure
              :op
-             :equality_constraints
-             :inequality_constraints
```
Comment (YingboMa) [inline src/systems/abstractsystem.jl:183]: `:constraints` should be here.

File: src/systems/optimization/optimizationsystem.jl [hunk 3]
```diff
@@ -124,6 +126,25 @@ function equations(sys::OptimizationSystem)
...
+end
+# namespace_constraint(ineq::Inequality, sys) = Inequality(
+#     namespace_expr(ineq.lhs, sys), 
+#     namespace_expr(ineq.rhs, sys),
+#     ineq.relational_op,
+# )
+function namespace_constraints(sys::OptimizationSystem)
+    namespace_constraint.(get_constraints(sys), Ref(sys))
+end
+get_constraints(sys::OptimizationSystem) = value(sys.constraints)
```
Comment (YingboMa) [inline src/systems/optimization/optimizationsystem.jl:141]: `sys.x` means getting the variable named `x` from `sys`.

**Review by YingboMa — CHANGES_REQUESTED** [review]
`sys.constraints` is deprecated.

File: src/systems/optimization/optimizationsystem.jl [hunk 4]
```diff
@@ -119,10 +121,38 @@ function generate_function(sys::OptimizationSystem, vs = states(sys), ps = param
...
+    if isempty(systems) 
+        op
+    else
+        op + reduce(+, map(sys_ -> namespace_expr.(get_op(sys_), sys_), systems))
+    end
+end
+
+namespace_constraint(eq::Equation, sys) = namespace_equation(eq, sys)
+
+# namespace_constraint(ineq::Inequality, sys) = namespace_inequality(ineq, sys)
```
Comment (YingboMa) [inline src/systems/optimization/optimizationsystem.jl:135]: Why are they removed?

Comment (ValentinKaisermayer) [reply src/systems/optimization/optimizationsystem.jl:135]: They are not removed but rather I put them in and commented them out. Needs https://github.com/JuliaSymbolics/Symbolics.jl/pull/717 first

**Comment by YingboMa** [comment]
The CI failure looks real.

**Comment by YingboMa** [comment]
Format

**Comment by ValentinKaisermayer** [comment]
I swear I did that:
```julia
(@v1.8) pkg> st
Status .julia\environments\v1.8\Project.toml`
  [70b36510] AutomaticDocstrings v1.0.5
  [98e50ef6] JuliaFormatter v1.0.9
  [295af30f] Revise v3.4.0

julia> format(".")
true
```

**Comment by ChrisRackauckas** [comment]
using the SciML format?

**Comment by ValentinKaisermayer** [comment]
Isn't that picked up automatically from the `.JuliaFormatter.toml`?

**Comment by ValentinKaisermayer** [comment]
Just to make sure
```
julia> JuliaFormatter.format("."; style=JuliaFormatter.SciMLStyle)
true
```

**Comment by YingboMa** [comment]
The format failure is due to https://github.com/SciML/ModelingToolkit.jl/commit/56ce7e5aae0e72786cf6666532173d5f455f8d3f . @baggepinnen please don't commit on master directly, and if you did it by mistake, please revert it if the CI fails.

**Comment by baggepinnen** [comment]
Sorry about that :( I was trying to figure out how I managed to commit straight to master, and I think it's because edits in the web interface defaults to committing instead of opening a PR
image
is there some way to make opening the PR the default option, like a branch-protection rule? I hadn't even noticed that I already committed :/

**Pull request merged**